### ISSUE TITLE
`AnyUrl` requires host for validation but `.host` is typed `None | str` 

### ISSUE THREAD

**Issue by mfontanaar** [pydantic/pydantic#9523]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

As far as I understand it `url = AnyUrl("")` can't be constructed if a host (and scheme) is not provided. Even more, after initialization, `url.host` can't be overridden neither. However, if I expect the type signature for `url.host` it shows `None | str`. Looking at [`_pydantic_core.pyi`](https://github.com/pydantic/pydantic-core/blob/fd262933683fad0a9706e63298065f0d6ab9905d/python/pydantic_core/_pydantic_core.pyi#L511), this seems accurate.

I think this could be a bug in the typing. This leads my code to either skipping static checks for the line, or having untestable codepaths. Consider the following example.

```python
def extract_host_and_path_1(url: AnyUrl) -> tuple[str, str]:
    if not url.path:
        raise Exception()
    return url.host, url.path
    

def extract_host_and_path_2(url: AnyUrl) -> tuple[str, str]:
    if not url.host:
        raise Exception()
    if not url.path:
        raise Exception()
    return url.host, url.path
```
The first example fails static checking. The second one is untestable, as I cannot build an `AnyUrl` instance for which `url.host` is None. For the time being I'm subclassing and testing with a custom class, but it seems like an overkill.

Best regards,

### Python, Pydantic & OS Version

```Text
pydantic version: 2.7.1
        pydantic-core version: 2.18.2
          pydantic-core build: profile=release pgo=true
                 install path: /home/user/.cache/pypoetry/virtualenvs/myproj-YzuwbEUn-py3.11/lib/python3.11/site-packages/pydantic
               python version: 3.11.7 (main, Dec 18 2023, 00:00:00) [GCC 13.2.1 20231205 (Red Hat 13.2.1-6)]
                     platform: Linux-6.7.3-200.fc39.x86_64-x86_64-with-glibc2.38
             related packages: pydantic-settings-2.2.1 typing_extensions-4.12.0
                       commit: unknown
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
@mfontanaar,

Agreed, this is either a bug or the docs are wrong. I think we should probably remote the note in the docs about the TLD.

**Comment by sydney-runkle** [comment]
Yeah, I can confirm this is a typing issue. I'll fix this tomorrow.

**Comment by sydney-runkle** [comment]
So this is a bit more complicated than anticipated, but I'll fix this with https://github.com/pydantic/pydantic/issues/7353
